### DISCUSSION TITLE
Bug Report: inconsistent query handling at rewrites

### DISCUSSION THREAD

**Discussion by ivan-kleshnin** [vercel/vercel#4389] [General] [answered]
# Bug report

### Why here?

This form is a placeholder: ![report-form](https://user-images.githubusercontent.com/2128182/82025926-aeccaa80-969a-11ea-93ab-1cd94a80d139.png) No attachments, nothing. So I'll report here this time.

## Bug Description

The following NextJS rewrite:

```js
module.exports = {
  experimental: {
    rewrites() {
      return [
        {
          source: "/content/:path*",
          destination: "/api/content?path=:path*",
        },
      ]
    },
  },
}
```

yields **one query format** with Now Dev

![dev](https://user-images.githubusercontent.com/2128182/82026782-e425c800-969b-11ea-9a6e-6992bb40ef4f.png)

and yields **another query format** on Zeit Platform

![prod](https://user-images.githubusercontent.com/2128182/82026203-17b42280-969b-11ea-9044-301b7ffa1f11.png)

```
{path: "foo.txt"} -- now dev

vs

{path: ["/foo.txt"]} -- production
```

*Please consider that to discover such bugs one has to deploy multiple times (wasting precious lambda count)* 😢

**Comment by leo** [comment]
Thank you for sharing this! We're currently setting up a integration test suite for ensuring production and development never diverge. We'll include this fix and let you know.

**Comment by ijjk** [comment, marked as answer]
Hi, thanks for the report! It looks like this behavior was incorrect in Next.js' server and was corrected in [this PR](https://github.com/zeit/next.js/pull/12930) which is available in `v9.4.1-canary.6` of Next.js. 

Multi-match params can be expected to always be arrays and then normal params should always be strings. If you notice any other inconsistencies please let us now!

**Comment by ivan-kleshnin** [comment]
Thanks for the quick response guys. Unfortunately I still observe the same behavior with `next dev` / `now dev` having `v9.4.1-canary.6` and `v9.4.1-canary.7` installed.

For the following setup

```
{
  source: "/content/:path*",
  destination: "/api/content?path=:path*",
},
```

```
GET content/foo.txt
```

resolves to 

```
req.query == { path: 'foo.txt' }
``` 

while

```
GET content/foo/bar.txt
```

yields 

```
req.query == { path: 'foobar.txt' }
```

**Reply by ijjk** [reply]
I see, thanks for the additional info! 

In this case it looks like the `path` query value should be a string since it is compiled and should be `foo/bar.txt`. The array value for multi-match params is only for the automatically added query values and you are overriding it here. It does look like the incorrect value is being compiled when manually defining the query value which should be addressed by [this PR](https://github.com/zeit/next.js/pull/13017)

Note: since the query values are automatically added to the query you do not need to add them manually like this.

**Reply by ivan-kleshnin** [reply]
For the potential readers – the above is correct.
As I removed explicit `path` argument I saw the expected array.
Thank you for the clarification @ijjk!
